msal/src/main/java/com/microsoft/identity/client/AccountAdapter.java

```diff
@@ -1,125 +1,296 @@
+                                                      @NonNull final List<ICacheRecord> homeRecords) {
+            // Init our sought value
+            final String guestAccountHomeAccountId = guestRecord.getAccount().getHomeAccountId();
+
+            // Create a List of home_account_ids from the homeRecords...
+            final List<String> homeAccountIds = new ArrayList<String>() {{
+                for (final ICacheRecord cacheRecord : homeRecords) {
+                    add(cacheRecord.getAccount().getHomeAccountId());
+                }
             }};
+
+            return homeAccountIds.contains(guestAccountHomeAccountId);
         }
 
-        accountOut.setAccountIdentifier(accountId);
-        accountOut.setHomeAccountIdentifier(homeAccountId);
-        accountOut.setUsername(accountIn.getUsername());
-        accountOut.setEnvironment(accountIn.getEnvironment());
+        @Override
+        public List<ICacheRecord> filter(@NonNull final List<ICacheRecord> records) {
+            final List<ICacheRecord> result = new ArrayList<>();
 
-        Logger.verbosePII(
-                TAG + methodName,
-                "Username: [" + accountIn.getUsername() + "]"
+            // First, get the home accounts....
+            final List<ICacheRecord> homeRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(false)
+                    );
+
+            // Then, get the guest accounts...
+            final List<ICacheRecord> guestRecords =
+                    filterCacheRecords(
+                            records,
+                            new HomeAccountFilter(true)
+                    );
+
+            // Iterate over the guest accounts and find those which have no associated home account
+            for (final ICacheRecord guestRecord : guestRecords) {
+                if (hasNoCorrespondingHomeAccount(guestRecord, homeRecords)) {
+                    result.add(guestRecord);
+                }
+            }
+
+            return result;
+        }
+    };
+
+    /**
+     * For a supplied List of ICacheRecords, create each root IAccount based on the home
+     * account and then add child-nodes based on any authorized tenants.
+     *
+     * @param allCacheRecords
+     * @return
+     */
+    @NonNull
+    static List<IAccount> adapt(@NonNull final List<ICacheRecord> allCacheRecords) {
+        // First, get all of the ICacheRecords for home accounts...
+        final List<ICacheRecord> homeCacheRecords = filterCacheRecords(
+                allCacheRecords,
+                new HomeAccountFilter(false)
         );
 
-        Logger.verbosePII(
-                TAG + methodName,
-                "Environment: [" + accountIn.getEnvironment() + "]"
+        // Then, get all of the guest accounts...
+        final List<ICacheRecord> guestCacheRecords = filterCacheRecords(
```

Comment:
nit: I'd explicitly specify that guestCacheRecords and guestCacheRecordsWithNoHomeAccount are mutually exclusive - not a subset of the other.

Reply:
I've clarified this via additional comments in 9188954824092606374d0523b66a25839bc1de9e - let me know if you'd like to see anything else here. Thanks @rpdome